vcd/resource_vcd_rde.go

```diff
@@ -0,0 +1,437 @@
+				Description: "The organization of the Runtime Defined Entity",
+				Computed:    true,
+			},
+			"resolve": {
+				Type: schema.TypeBool,
+				Description: "If `true`, the Runtime Defined Entity will be resolved by this provider. If `false`, it won't be" +
+					"resolved and must be either done by an external component or with an update. The Runtime Defined Entity can't be" +
+					"deleted until the entity is resolved.",
+				Required: true,
+			},
+			"resolve_on_removal": {
+				Type: schema.TypeBool,
+				Description: "If `true`, the Runtime Defined Entity will be resolved before it gets deleted, to forcefully delete it." +
+					"Otherwise, destroy will fail if it is not resolved.",
+				Default:  false,
+				Optional: true,
+			},
+			"state": {
+				Type: schema.TypeString,
+				Description: "When created it will be in PRE_CREATED state. If the specified JSON in either `input_entity` or `input_entity_url` is correct, the state will be RESOLVED," +
+					"otherwise it will be RESOLUTION_ERROR. If an entity in an RESOLUTION_ERROR state, it will require to be updated to a correct JSON to be usable",
+				Computed: true,
+			},
+			"entity_in_sync": {
+				Type:        schema.TypeBool,
+				Description: "If true, `computed_entity` is equal to either `input_entity` or the contents of `input_entity_url`",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func resourceVcdRdeCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	name := d.Get("name").(string)
+	rdeTypeId := d.Get("rde_type_id").(string)
+
+	rdeType, err := vcdClient.GetRdeTypeById(rdeTypeId)
+	if err != nil {
+		return diag.Errorf("could not create RDE with name '%s', could not retrieve RDE Type with ID '%s': %s", name, rdeTypeId, err)
+	}
+
+	tenantContext := govcd.TenantContext{}
+	if vcdClient.Client.IsSysAdmin {
+		org, err := vcdClient.GetAdminOrgFromResource(d)
+		if err != nil {
+			return diag.Errorf("could not create RDE with name '%s', error retrieving Org '%s': %s", name, d.Get("org").(string), err)
+		}
+		tenantContext.OrgId = org.AdminOrg.ID
+		tenantContext.OrgName = org.AdminOrg.Name
+	}
+
+	jsonSchema, err := getRdeJson(vcdClient, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	rde, err := rdeType.CreateRde(types.DefinedEntity{
+		Name:       name,
+		ExternalId: d.Get("external_id").(string),
+		Entity:     jsonSchema,
+	}, &tenantContext)
+	if err != nil {
+		return diag.Errorf("could not create the Runtime Defined Entity '%s' of type '%s': %s", name, rdeTypeId, err)
+	}
+
+	// We save the ID immediately as the Resolve operation can fail, but the RDE is already created. If this happens,
+	// it should go to the Update operation instead.
+	d.SetId(rde.DefinedEntity.ID)
+
+	if d.Get("resolve").(bool) {
+		err = rde.Resolve()
+		if err != nil {
+			return diag.Errorf("could not resolve the Runtime Defined Entity '%s' of type '%s': %s", name, rdeTypeId, err)
```

Comment:
Then lines 144 to 146 are false, right? My assumption is wrong.

Reply:
It is good to store `ID` (line 146) as otherwise we would have an orphan resource (nothing would be stored in state, while the entity would be created in VCD), but when we have `d.SetId` and then return non `nil` error - Terraform will automatically `taint` that resource so that it is recreated on next step

Reply:
Got it, thanks! I was thinking of improving the error message, but I guess that it's clear enough (if the docs were read)